language/vm/src/deserializer.rs

```diff
@@ -1510,3 +1517,219 @@ impl Opcodes {
+    // check and return early for invalid indicies
+    if script.clone().freeze().is_err() {
+        return Ok(());
+    }
+
+    let parameters_signature = &script.signatures[script.parameters.0 as usize];
+    let locals_idx = script.code.locals;
+    let locals_signature = &script.signatures[locals_idx.0 as usize];
+
+    // Find the parameters that need to be updated
+    let num_parameters = parameters_signature.len() as LocalIndex;
+    let signer_ref_parameters = parameters_signature
+        .0
+        .iter()
+        .enumerate()
+        .filter(|(_idx, sig_token)| {
+            matches!(
+                sig_token,
+                SignatureToken::Reference(inner) if matches!(&**inner, SignatureToken::Signer)
+            )
+        })
+        .map(|(idx, _sig)| idx as LocalIndex)
+        .collect::<Vec<_>>();
+    if signer_ref_parameters.is_empty() {
+        return Ok(());
+    }
+
+    // Make a new parameter signature
+    let new_parameters = Signature(
+        parameters_signature
+            .0
+            .iter()
+            .map(|s| match s {
+                SignatureToken::Reference(inner) if matches!(&**inner, SignatureToken::Signer) => {
+                    SignatureToken::Signer
+                }
+                // Map signer to signer ref to keep invalid scripts invalid
+                SignatureToken::Signer => {
+                    SignatureToken::Reference(Box::new(SignatureToken::Signer))
+                }
+                _ => s.clone(),
+            })
+            .collect::<Vec<_>>(),
+    );
+
+    // Make a new local signature
+    let (new_locals_signature, new_local_idxs) = {
+        let mut new_sig = (0..signer_ref_parameters.len())
+            .map(|_| SignatureToken::Reference(Box::new(SignatureToken::Signer)))
+            .collect::<Vec<_>>();
+        new_sig.extend(locals_signature.0.iter().cloned());
+        let new_local_idxs = (0..signer_ref_parameters.len())
+            .map(|idx| idx as LocalIndex + num_parameters)
+            .collect::<Vec<_>>();
+        (Signature(new_sig), new_local_idxs)
+    };
+
+    // Determine a mapping from the owned signers to the new locals with the borrowed signer param
+    // Build instructions to borrow those params
+    let (signer_ref_remapping, mut borrow_owned_signers_instructions) = {
+        let mut new_code = vec![];
+        let mut signer_ref_remapping = (0..num_parameters).map(|_| None).collect::<Vec<_>>();
+
+        for (signer_ref_parameter_idx, new_local_idx) in
+            signer_ref_parameters.iter().zip(&new_local_idxs)
+        {
+            signer_ref_remapping[*signer_ref_parameter_idx as usize] = Some(new_local_idx);
+            new_code.push(Bytecode::ImmBorrowLoc(*signer_ref_parameter_idx));
+            new_code.push(Bytecode::StLoc(*new_local_idx));
+        }
+        (signer_ref_remapping, new_code)
+    };
+
+    // Update the code
```

Comment:
I see that there's no way around doing this. But I am also concerned that this transformation changes the gas cost of a script... Of course, gas cost can also change because we are allowed to adjust the per-instruction costs, so maybe this is no different. But definitely something we should call out in the changelog + account for in our gas cost estimates for old tx scripts. CC @tzakian .

Reply:
Ah yes it does, which might be tricky...